Code/GraphMol/DetermineBonds/DetermineBonds.cpp

```diff
@@ -18,9 +18,104 @@
 #include <unordered_map>
 #include <boost/graph/adjacency_list.hpp>
 #include <boost/graph/max_cardinality_matching.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
 
 typedef boost::adjacency_list<boost::vecS, boost::vecS, boost::undirectedS>
     Graph;
+using boost::multiprecision::uint1024_t;
+
+namespace {
+
+
+// see http://phrogz.net/lazy-cartesian-product
+template <typename T>
+struct LazyCartesianProduct {
+  std::vector<std::vector<T>> listOfLists;
+  std::vector<uint1024_t> divs;
+  std::vector<uint1024_t> mods;
+  uint1024_t max_size;
+  uint1024_t currentPos;
+
+  explicit LazyCartesianProduct(std::vector<std::vector<T>> &input)
+      : listOfLists(input), currentPos(0) {
+    unsigned int size = listOfLists.size();
+    divs.resize(size);
+    mods.resize(size);
+    max_size = 1;
+
+    for (int i = size - 1; i >= 0; --i) {
+      uint1024_t items(listOfLists[i].size());
+      divs[i] = max_size;
+      mods[i] = items;
+      max_size *= items;
+    }
+  }
+
+  std::vector<T> entry_at(uint1024_t pos) const;
+  std::vector<T> next() { return entry_at(currentPos++); }
+  bool atEnd() const { return currentPos >= max_size; }
+};
+
+template <typename T>
+std::vector<T> LazyCartesianProduct<T>::entry_at(uint1024_t pos) const {
+  auto length = listOfLists.size();
+  std::vector<T> res(length);
+  for (auto i = 0; i < length; ++i) {
+    res[i] = listOfLists[i][(unsigned long long)((uint1024_t)(pos / divs[i]) %
+                                                 mods[i])];
```

Comment:
We avoid c-style casts whenever possible
```suggestion
    res[i] = listOfLists[i][static_cast<size_t>(static_cast<uint1024_t>(pos / divs[i]) %
                                                 mods[i])];
```